libs/x-content/src/main/java/org/elasticsearch/xcontent/support/filtering/FilterPath.java

```diff
@@ -11,102 +11,194 @@
-    }
+            BuildNode(boolean isFinalNode) {
+                children = new HashMap<>();
+                this.isFinalNode = isFinalNode;
+            }
+        }
 
-    public static FilterPath[] compile(Set<String> filters) {
-        if (filters == null || filters.isEmpty()) {
-            return null;
+        private BuildNode root = new BuildNode(false);
+
+        void insert(String filter) {
+            insertNode(filter, root);
         }
 
-        List<FilterPath> paths = new ArrayList<>();
-        for (String filter : filters) {
-            if (filter != null) {
-                filter = filter.trim();
-                if (filter.length() > 0) {
-                    paths.add(parse(filter, filter));
+        FilterPath build() {
+            return buildPath("", root);
+        }
+
+        void insertNode(String filter, BuildNode node) {
+            int end = filter.length();
+            for (int i = 0; i < end;) {
+                char c = filter.charAt(i);
+                if (c == '.') {
+                    String field = filter.substring(0, i).replaceAll("\\\\.", ".");
+                    BuildNode child = node.children.get(field);
+                    if (child == null) {
+                        child = new BuildNode(false);
+                        node.children.put(field, child);
+                    }
+                    if (false == child.isFinalNode) {
+                        insertNode(filter.substring(i + 1), child);
+                    }
+                    return;
+                }
+                ++i;
+                if ((c == '\\') && (i < end) && (filter.charAt(i) == '.')) {
+                    ++i;
                 }
             }
+
+            String field = filter.replaceAll("\\\\.", ".");
+            node.children.put(field, new BuildNode(true));
         }
-        return paths.toArray(new FilterPath[paths.size()]);
-    }
 
-    private static FilterPath parse(final String filter, final String segment) {
-        int end = segment.length();
+        FilterPath buildPath(String segment, BuildNode node) {
+            Map<String, FilterPath> termsChildren = new HashMap<>();
+            List<FilterPath> wildcardChildren = new ArrayList<>();
+            for (Map.Entry<String, BuildNode> entry : node.children.entrySet()) {
+                String childName = entry.getKey();
+                BuildNode childNode = entry.getValue();
+                FilterPath childFilterPath = buildPath(childName, childNode);
+                if (childName.contains(WILDCARD)) {
+                    wildcardChildren.add(childFilterPath);
+                } else {
+                    termsChildren.put(childName, childFilterPath);
+                }
+            }
+
+            boolean doubleWildcard = segment.equals(DOUBLE_WILDCARD);
+            boolean isFinalNode = node.isFinalNode;
+            boolean hasDoubleWildcard = hasDoubleWildcard(segment, termsChildren, wildcardChildren);
```

Comment:
I think we can remove this precalculation. We only need it when `exclude`ing and that's not super widely used.